Feat5352  New WixBA UI

phillHgl opened the pull request:
Submitting a 'functional' revised WixBA for review and more feedback.
-   I researched Windows UI guidelines, Windows 10 behavior, and compared to the AppX install demonstrated in the online meeting #109.  I tried to implement the feedback, ultimately making sure that I authored to use the OS default behavior (removing the custom styles that I had added to emulate the AppX screen shots as there were too many conflicts with the guidelines and Win 10 behavior.  Still interested in feedback on these issues.
-   Implemented display of Update details from the atom.  Not having access to upload a new build, I 'mocked' the build to target the WiX 3.10 atom feed for prototyping.  The HTML content is feed to a WebBrowser object, notorious of bad behavior especially in regard to changing the WPF layout when resizing.   I researched other approaches such as using a  FlowDocument, but decided to set Min and Max on the resize range to try and hide the bad behavior.
-  Implemented resizing (painfully) but I think functionally.  I think that the CanMinimize option is more maintainable.
-  The targeted framework remains NetFx40, which means I did not find a way to detect if the theme is changed while the app is running.  Otherwise the high contrast themes and screed reader accessibility tools are supported (for those of us that prefer flashier displays).
-  I have only tested each of the setup phases, as best I could (with a simplified PackageGroups.wxs) on Win 10 x64, but this should allow others to also try it out.

phillHgl committed:
Start work on a facelift, with IVM hacks to see the layout of normally hidden buttons

phillHgl committed:
implemented progress and status, except final status

phillHgl committed:
implemented facelift, test core install, repair, uninstall

phillHgl committed:
add files not added in past commit

phillHgl committed:
removed old resources and updated styles and title layout

phillHgl committed:
added update button, oops!

phillHgl committed:
Merge remote-tracking branch 'upstream/develop' into WixBAFacelift

phillHgl committed:
added WPF Behavior to address WIXBUG #5264 using sample code found on SO, as credited, but removed the DragMoveBehavior as it was throwing an exception and our concern is addres with the NoSnapBehavior.

phillHgl committed:
changes to style as creating screen shots

phillHgl committed:
refactor styles

phillHgl committed:
refactor Title and SKU areas

phillHgl committed:
action button styles

phillHgl committed:
update info layout

phillHgl committed:
refactor display html, with WiX Mocks

phillHgl committed:
resizing is best that can be done at this time.

phillHgl committed:
implementd display package name rather than Action Text

phillHgl committed:
testing, reverted MOCK

phillHgl committed:
hide icon, refactor title and action area for resizing

phillHgl committed:
implemented brightness detect for theme controls since not using NetFx45 events.

phillHgl committed:
clean up styles and set Tab stops

phillHgl committed:
added history, removed extra automation labels and reviewed comments

robmen commented on src/Setup/WixBA/InstallationViewModel.cs:
```diff
@@ -254,6 +277,33 @@ public bool UninstallEnabled
             get { return this.UninstallCommand.CanExecute(this); }
         }
 
+        private ICommand openLogCommand;
```
Style: We stick all privates at the top of the class.

phillHgl replied:
Agreed.  I knew I put it there and when I went back to make that change I
missed it.

On Thu, Jul 28, 2016 at 3:04 AM, Rob Mensching notifications@github.com
wrote:

> In src/Setup/WixBA/InstallationViewModel.cs
> https://github.com/wixtoolset/wix4/pull/208#discussion_r72579723:
> 
> > @@ -254,6 +277,33 @@ public bool UninstallEnabled
> >              get { return this.UninstallCommand.CanExecute(this); }
> >          }
> > -        private ICommand openLogCommand;
> 
> Style: We stick all privates at the top of the class.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/wixtoolset/wix4/pull/208/files/5b08dba511d7438508767effab67896442e4ad1b#r72579723,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AHyCLUkk7Bi6zvlPyOGuB99HouPFbrUCks5qaGJ9gaJpZM4JVm8P
> .

robmen commented on src/Setup/WixBA/WixBA.cs:
```diff
@@ -1,5 +1,7 @@
 // Copyright (c) .NET Foundation and contributors. All rights reserved. Licensed under the Microsoft Reciprocal License. See LICENSE.TXT file in the project root for full license information.
 
+// WixBAFacelift TODO list:
+// Compile against NetFx452 (supported).  Moved to Net45 to check out using WindowChrome, then backed out Chrome.  check with wix-devs, then update Bundle when framework consenuse exist.
```
Do we have to? My hope is that if we find a NETFX v4.0 that we'll be happy to use that. Do we need v4.5+?

phillHgl replied:
Sorry I should have removed that comment.  It was just a reminder to have
further discussion pointing out that I tried to use the Chrome object, but
backed it out because we want to compile against NetFx40. I don't recall
why folks indicated that the Chrome object was a better path, but I don't
see any problem staying with NetFx40.

On Thu, Jul 28, 2016 at 3:09 AM, Rob Mensching notifications@github.com
wrote:

> In src/Setup/WixBA/WixBA.cs
> https://github.com/wixtoolset/wix4/pull/208#discussion_r72580324:
> 
> > @@ -1,5 +1,7 @@
> >  // Copyright (c) .NET Foundation and contributors. All rights reserved. Licensed under the Microsoft Reciprocal License. See LICENSE.TXT file in the project root for full license information.
> > 
> > +// WixBAFacelift TODO list:
> > +// Compile against NetFx452 (supported).  Moved to Net45 to check out using WindowChrome, then backed out Chrome.  check with wix-devs, then update Bundle when framework consenuse exist.
> 
> Do we have to? My hope is that if we find a NETFX v4.0 that we'll be happy
> to use that. Do we need v4.5+?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/wixtoolset/wix4/pull/208/files/5b08dba511d7438508767effab67896442e4ad1b#r72580324,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AHyCLaCuslwW6U1ISG13rYxIp2oi3ugsks5qaGPLgaJpZM4JVm8P
> .

robmen replied:
Awesome.

rseanhall replied:
Please remove these comments.

robmen commented on src/Setup/WixBA/WixBA.cs:
```diff
@@ -61,6 +63,65 @@ public static void LaunchUrl(string uri)
+        internal static void OpenLog(Uri uri)
+        {
+            // Switch the wait cursor since shellexec can take a second or so.
+            System.Windows.Input.Cursor cursor = WixBA.View.Cursor;
```
The bulk of this should be refactored to be a reusable `ShellExecuteWithBetterCursorAndBetterNameThanRobCameUpWithHere()` method.

phillHgl replied:
Not sure that I understand but I will take a look at it tomorrow.

On Thu, Jul 28, 2016 at 3:10 AM, Rob Mensching notifications@github.com
wrote:

> In src/Setup/WixBA/WixBA.cs
> https://github.com/wixtoolset/wix4/pull/208#discussion_r72580457:
> 
> > @@ -61,6 +63,65 @@ public static void LaunchUrl(string uri)
> >          }
> > 
> > ```
> >      /// <summary>
> > ```
> > -        /// Open a log file.
> > -        /// </summary>
> >   
> >   
> > -        /// <param name="uri">URI to a log file.</param>
> > -        internal static void OpenLog(Uri uri)
> > -        {
> > -            // Switch the wait cursor since shellexec can take a second or so.
> > -            System.Windows.Input.Cursor cursor = WixBA.View.Cursor;
> 
> The bulk of this should be refactored to be a reusable
> ShellExecuteWithBetterCursorAndBetterNameThanRobCameUpWithHere() method.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/wixtoolset/wix4/pull/208/files/5b08dba511d7438508767effab67896442e4ad1b#r72580457,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AHyCLVcwB-JRSi8RMbQ6RccYgnq50ISdks5qaGQNgaJpZM4JVm8P
> .

robmen replied:
I'm basically just saying this method and the one that follows can be refactored to share a private `ShellExecute` function to reduce code duplication.

phillHgl replied:
Understood. Will do.  Procedurally I did not view this PR as a 'complete' feat5352 but as you requested that I submit 'functional' code so that others could try it out and we could do more refactoring.  I already started simplifying some of the XAML which has been my primary focus rather than the backing VM code.  I also found a VS extension  XamlStyler.Package.vsix) which I plan to try out as I know that I use a flat XAML syle different from what I assume most folks prefer.  So my procedure question is do I just make these changes and submit another PR or how do you want the PR and these changes integrated?

robmen commented on src/Setup/WixBA/WixBA.csproj:
```diff
@@ -1,19 +1,50 @@
     <AssemblyName>WixBA</AssemblyName>
     <OutputType>Library</OutputType>
     <RootNamespace>WixToolset.UX</RootNamespace>
+    <TargetFrameworkVersion>v4.0</TargetFrameworkVersion>
```
This whole block should be unnecessary and come from the `tools\WixBuild.csproj.props`, right?

phillHgl replied:
I really don't know.  The only reason that the PR had a diff here is because I experimented with compiling to NetFx45 and then after the online meeting 109 I reverted back to NetFx40 as requested.  I will look at  tools\WixBuild.csproj.props and remove this block if that seems wise.

Is your comment just about TargetFrameworkVersion? I see it defined in tools\WixBuild.csproj.props so it does seem unnecessary to me.  But if your comment is related to RootNamespace or AssemblyName then I am confused.  I will assume a narrow intent until clarified.

robmen commented on src/common/WixDistribution.cs:
```diff
@@ -34,6 +34,11 @@ internal static class WixDistribution
         /// <summary>
+        /// Product name for the distribution.
+        /// </summary>
+        public static string Product = "Windows Installer XML Toolset";
```
We're not called "Windows Installer XML Toolset" any more (we should fix places that use this old name). It's just "WiX Toolset" now.

phillHgl replied:
OK I will change it.

On Thu, Jul 28, 2016 at 3:20 AM, Rob Mensching notifications@github.com
wrote:

> In src/common/WixDistribution.cs
> https://github.com/wixtoolset/wix4/pull/208#discussion_r72581636:
> 
> > @@ -34,6 +34,11 @@ internal static class WixDistribution
> >          public static string ShortProduct = "WiX Toolset";
> > 
> > ```
> >      /// <summary>
> > ```
> > -        /// Product name for the distribution.
> > -        /// </summary>
> >   
> >   
> > -        public static string Product = "Windows Installer XML Toolset";
> 
> We're not called "Windows Installer XML Toolset" any more (we should fix
> places that use this old name). It's just "WiX Toolset" now.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/wixtoolset/wix4/pull/208/files/5b08dba511d7438508767effab67896442e4ad1b#r72581636,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AHyCLUkLxskcR2Tco3jRGZtfIOZUlGoNks5qaGZMgaJpZM4JVm8P
> .

robmen replied:
Just use `ShortProduct`. :)

robmen commented:
What is `src/Setup/WixBA/Resources/wix-light-san - Copy.ico` for?

phillHgl committed:
added ZIndex to avoid overlapped browser control hiding action text updates

phillHgl committed:
clear the message to avoid a stale display

phillHgl commented:
In the online meeting I pointed out that when a Window is created in WPF
with the chrome style of resizable or even CanMinimize, it puts a icon in
the upper left corner.  There is no option to not have the icon.

https://msdn.microsoft.com/en-us/library/system.windows.window.icon(v=vs.110).aspx
From the above link:
"A WPF window always displays an icon. When one is not provided by setting
Icon, WPF chooses an icon to display based on the following rules:

   1.

   Use the assembly icon, if specified.
   2.

   If the assembly icon is not specified, use the default Microsoft Windows
   icon.

"

So as I mention in the 109 meeting, since folks wanted a 'Win 10' dialog
without the icon, one approach was to make and icon which is just simply
transparent so that the color of the chorme area shows through regardless
of the system theme.  The negative of this is that on older OS Alt-Tab
displays a list of icons, which in this case does not exist.  On Win 10 it
displays a thumbprint of the main window.  So let me know what you want
done on this issue.

On Thu, Jul 28, 2016 at 3:25 AM, Rob Mensching notifications@github.com
wrote:

> What is src/Setup/WixBA/Resources/wix-light-san - Copy.ico for?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/wixtoolset/wix4/pull/208#issuecomment-235831537, or mute
> the thread
> https://github.com/notifications/unsubscribe-auth/AHyCLY3EReu8UpMa5JWYCR1ZYrr_LGV2ks5qaGdfgaJpZM4JVm8P
> .

robmen commented:
:shrug: Low priority. Let's just use an icon for now and we can tackle it in future if it really bugs us.

phillHgl commented:
As for my procedural question above I see that by adding commits they get added in, so that answers my questions.  I will work on these issues. Thanks for the feedback.

phillHgl committed:
refactor functions used to shell to logs and urls, correct IVM style issues, refactor view to use TextBlock rather than ContentPresenter where appropriate.

phillHgl commented:
> > What is  src/Setup/WixBA/Resources/wix-light-san - Copy.ico  for?
> > Actually that file should have been removed along with:
> > src/Setup/WixBA/Resources/wix-light-san.ico

Those were left over from an experiment that I tried to edit an existing product's icon to get the transparent icon.  So those two files should have been removed and not checked in at all.  However as submitted the PR used san-image.ico as a transparent icon file in order to hide the default icon that the WPF framework assigns to any window.  Based on you latest comment I also removed san-image.ico so now the default icon is displayed.  If you want a particular icon displayed, it can be added at any point.

I'm just testing these changes and then I will commit changes for all off the feedback to date.

phillHgl committed:
implemented PR feedback on all points.

DoCode commented:
@phillHgl how can I build your WixBA and test/see the UI?

phillHgl commented:
There is a link on how to build the wixtoolset at this page:
http://wixtoolset.org/development/

The Setup project is one of the projects,which is built using the wix4
build tools.  Last week I also posted some more details in the wix-users
forum, but I don't have access to a specific link at this time.  You can
search the wix-users forum archive for that post. Thanks for the interest.

On Fri, Jul 29, 2016 at 2:01 PM, DoCode notifications@github.com wrote:

> @phillHgl https://github.com/phillHgl how can I build your WixBA and
> test/see the UI?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/wixtoolset/wix4/pull/208#issuecomment-236266082, or mute
> the thread
> https://github.com/notifications/unsubscribe-auth/AHyCLTZhUxmOlBIKLJKeFh9oVpxXoj0_ks5qak4AgaJpZM4JVm8P
> .

phillHgl committed:
mostly xaml style changes, but testing on remotevm still crashes, so commiting changes to switch to known good develop branch to test build/remote test process

phillHgl committed:
removed WixDistribution.Product, which was set to the 'old name' after not finding and instance of using it in the src

phillHgl commented:
I resolved my 'unofficial official' build concerns and tested all phases of install, repair, uninstall on a remote VM.  I also built a branch of the develop branch (without WixBA changes) and then 'upgraded' to a feat5352 build, which went well but since the auto-gened version is the same I will have to try this again latter to be certain of the results.   I believe I have addressed all feedback, but I am interested to know if there are any other style changes which are desired.  I plan to stand down on making further changes until I hear more specific feedback.

robmen commented:
I pulled locally and all seemed to work. Yay! :tada:

Only feedback I have is the title font looks funny (italics or something) and is centered. I think we should just make it left aligned and a bit bigger but same font. But that is minor tweak. This is looking great.

rseanhall commented:
This is looking really good.

When I launch it, the install button is disabled until I click somewhere.

rseanhall commented on src/Setup/WixBA/InstallationViewModel.cs:
```diff
@@ -502,15 +452,21 @@ private void CacheAcquireComplete(object sender, CacheAcquireCompleteEventArgs e
         private void ExecutePackageBegin(object sender, ExecutePackageBeginEventArgs e)
         {
-            this.executePackageStart = e.ShouldExecute ? DateTime.Now : DateTime.MinValue;
+            lock (this)
```
The burn engine is single threaded, there shouldn't be a need to lock here.

phillHgl replied:
From gmail I wrote a response to this, but since I don't see it, and at the risk of duplicating, let me state my concern with removing the lock().  As I understand it in WPF each ViewModel is on a separate thread, and long ago when my mba was having some issues, I noticed that the WixBA (3.8?) used lock() when more than one VM provides a handler for the same event.  Adding lock() in that scenario resolved my problems, though I do not recall the exact details.  If you want me to remove the lock() I can do so, but given that each VM is a separate thread, doesn't using the lock() serve a valid purpose here?  (Please take this as a comment for each of the places where a concern was raised about using lock(). )

rseanhall replied:
Never mind, not a big deal.  Each VM does not get its own thread and only one event handler will be running at a time, but better safe than sorry.

rseanhall commented on src/Setup/WixBA/InstallationViewModel.cs:
```diff
@@ -502,15 +452,21 @@ private void CacheAcquireComplete(object sender, CacheAcquireCompleteEventArgs e
         private void ExecutePackageComplete(object sender, ExecutePackageCompleteEventArgs e)
         {
-            if (DateTime.MinValue < this.executePackageStart)
+            lock (this)
```
or here

rseanhall commented on src/Setup/WixBA/NoSnapBehavior.cs:
```diff
@@ -0,0 +1,59 @@
+    /// the window from being snapped to the edge of the screen (AeroSnap).
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    public class NoSnapBehavior<T> : Behavior<T> where T : FrameworkElement
```
I thought we discussed that it should be resizable, and that the maximize/restore button would be available.  In that case, we shouldn't be fighting the Aero Snap.  Though this probably should be put in v3.

rseanhall commented on src/Setup/WixBA/ProgressViewModel.cs:
```diff
@@ -31,7 +31,8 @@ public ProgressViewModel(RootViewModel root)
             WixBA.Model.Bootstrapper.ExecutePackageBegin += this.ExecutePackageBegin;
-            WixBA.Model.Bootstrapper.ExecuteMsiMessage += this.ExecuteMsiMessage;
+            WixBA.Model.Bootstrapper.ExecutePackageComplete += this.ExecutePackageComplete;
+            //WixBA.Model.Bootstrapper.ExecuteMsiMessage += this.ExecuteMsiMessage;
```
Let's delete commented out code.

rseanhall commented on src/Setup/WixBA/ProgressViewModel.cs:
```diff
@@ -45,6 +46,7 @@ public bool ProgressEnabled
         {
             get { return this.root.InstallState == InstallationState.Applying; }
         }
+        
```
and delete extra lines.

rseanhall commented on src/Setup/WixBA/ProgressViewModel.cs:
```diff
@@ -127,22 +129,19 @@ private void PlanPackageComplete(object sender, PlanPackageCompleteEventArgs e)
-            this.Package = WixBA.Model.GetPackageName(e.PackageId);
-
-            e.Result = this.root.Canceled ? Result.Cancel : Result.Ok;
+            lock (this)
```
The lock shouldn't be necessary here.

rseanhall commented on src/Setup/WixBA/ProgressViewModel.cs:
```diff
@@ -127,22 +129,19 @@ private void PlanPackageComplete(object sender, PlanPackageCompleteEventArgs e)
-        private void ExecuteMsiMessage(object sender, ExecuteMsiMessageEventArgs e)
+        private void ExecutePackageComplete(object sender, ExecutePackageCompleteEventArgs e)
         {
             lock (this)
```
or here

rseanhall commented on src/Setup/WixBA/RootView.xaml:
```diff
@@ -1,125 +1,487 @@
 <?xml version="1.0" encoding="utf-8" ?>
-<!-- Copyright (c) .NET Foundation and contributors. All rights reserved. Licensed under the Microsoft Reciprocal License. See LICENSE.TXT file in the project root for full license information. -->
-
+<!--  Copyright (c) .NET Foundation and contributors. All rights reserved. Licensed under the Microsoft Reciprocal License. See LICENSE.TXT file in the project root for full license information.  -->
```
Why is there extra spaces added between `<!--` and `Copyright`?

rseanhall commented on src/Setup/WixBA/RootView.xaml:
```diff
@@ -1,125 +1,487 @@
+        xmlns:ux="clr-namespace:WixToolset.UX"
+        MinWidth="550"
+        MinHeight="400"
+        MaxWidth="1200"
```
If it's maximizable, there shouldn't be a MaxWidth or MaxHeight

phillHgl replied:
I think I pointed this out in the PR, but the issue is that we added a
WebBrowser control that is displayed when the UpdateViewModel detects an
available update.  The WPF WebBrowser control is notorious for having
layout problems, and from what I read there is no real solution to these
issues because it is a thin wrapper around a win32 com object.  So when
there was not min and max defined and the user was allowed to resize beyond
these ranges the WebBrowser Control on reduction gets pushed over the top
of other layout and on expansion above these limits it is less problematic
but also seemed to be rather ugly.

Yes it was proposed in the WIP review that we support user resizing and a
Max button.  I spent a lot of time on the above issue, and even tried for a
while to go down the path of eliminating the WebBrowser control and
implementing a solution using a FlowDocument, which was the recommendation
that I found when researching the problem.  That however requires importing
a HTML to FlowDocument conversion.  I was not clear on the license impacts
of the code that I found in this regard so I decided not to pursue that
path any further at this time.  And in the WIP review it was proposed that
maybe someone else would want to work on this aspect of the changes.

This is also the reason for the need for Areo No Snap.  The Max window
button is available, which maximizes to the MaxWidth which we defined so
that the above layout is not messed up.

The way that I tested this behavior, since I did not have access to staging
an available update, was to hack src/Setup/Bundle/Bundle.wxs to have a
lower version, such as 3.9.0.0, and then hack
src/common/WixDistribution.wxl to point at the 3.10 feed.  This allows the
bundle to detect a staged update and display the details.  I also used test
code in UpdateViewModel/DetectUpdates to insert more these into
this.UpdateChanges to test the layout with a large HTML string.

So I think there are competing goals here and I don't know how at this time
to please all perspectives.  Feel free to let me know your view or how
these issues should be prioritized.

On Sun, Jul 31, 2016 at 11:51 PM, Sean Hall notifications@github.com
wrote:

> In src/Setup/WixBA/RootView.xaml
> https://github.com/wixtoolset/wix4/pull/208#discussion_r72923919:
> 
> > -    xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
> > -    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
> > -    xmlns:ux="clr-namespace:WixToolset.UX"
> > -    WindowStartupLocation="CenterScreen"
> > -    WindowStyle="None"
> > -    AllowsTransparency="True"
> > -    Background="{x:Null}"
> > -    Width="400"
> > -    Height="400">
> > -        xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
> > -        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
> > -        xmlns:i="http://schemas.microsoft.com/expression/2010/interactivity"
> > -        xmlns:ux="clr-namespace:WixToolset.UX"
> > -        MinWidth="550"
> > -        MinHeight="400"
> > -        MaxWidth="1200"
> 
> If it's maximizable, there shouldn't be a MaxWidth or MaxHeight
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/wixtoolset/wix4/pull/208/files/28721ab8e4985a5641b3efd7975a67cecd8f8d4a#r72923919,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AHyCLbEPh3-BgixU82UqQpHMNoY3sRFzks5qbXtRgaJpZM4JVm8P
> .

rseanhall replied:
If the WebBrowser control is the issue, then let's put a max size on that instead of the window (or just make the WebBrowser a fixed size?).

phillHgl replied:
The WebBrowser control does have a MaxWidth value sit which is just
slightly smaller that the Window/@MaxWidth.  That is one of the few times
that I limit the width of a control but in this case when I mocked an
available update and added lots of 'content' HTML if made since to allow
the dialog to be almost the width of a sheet of paper (on my monitors).
But to allow the main window to continue to expand just looks strange.
Also as I re-listen to the 109 meeting, while demo-ing the Appx install Rob
noted that they also had set a Window MinWidth and I don't know if they had
a MaxWidth.  I will go and remove the MaxWidth (along with the Mouse
handling which is a different issue, and based on my testing that will
degrade the UX somewhat, but I am sure it will be acceptable to most).

On Mon, Aug 1, 2016 at 11:45 AM, Sean Hall notifications@github.com wrote:

> In src/Setup/WixBA/RootView.xaml
> https://github.com/wixtoolset/wix4/pull/208#discussion_r73011188:
> 
> > -    xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
> > -    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
> > -    xmlns:ux="clr-namespace:WixToolset.UX"
> > -    WindowStartupLocation="CenterScreen"
> > -    WindowStyle="None"
> > -    AllowsTransparency="True"
> > -    Background="{x:Null}"
> > -    Width="400"
> > -    Height="400">
> > -        xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
> > -        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
> > -        xmlns:i="http://schemas.microsoft.com/expression/2010/interactivity"
> > -        xmlns:ux="clr-namespace:WixToolset.UX"
> > -        MinWidth="550"
> > -        MinHeight="400"
> > -        MaxWidth="1200"
> 
> If the WebBrowser control is the issue, then let's put a max size on that
> instead of the window (or just make the WebBrowser a fixed size?).
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/wixtoolset/wix4/pull/208/files/28721ab8e4985a5641b3efd7975a67cecd8f8d4a#r73011188,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AHyCLQVsH5F1llGzqig1K8ZFVqnT96JTks5qbiKqgaJpZM4JVm8P
> .

rseanhall commented on src/Setup/WixBA/RootView.xaml:
```diff
@@ -1,125 +1,487 @@
+        <Border HorizontalAlignment="Stretch"
+                VerticalAlignment="Stretch"
+                Background="{DynamicResource {x:Static SystemColors.WindowBrushKey}}"
+                MouseLeftButtonDown="Background_MouseLeftButtonDown">
```
The old UI allowed dragging from clicking on the background since it didn't have the traditional window chrome.  We shouldn't be supporting this anymore.

phillHgl replied:
Sean, I disagree. I compiled builds with and without both this MouseLeftButtonDowna and the Mouse handlers in the following comment.  The UX behaviors is definitely improved by leaving this authoring as it is.  Yes we now have a window chrome, but as with the example AppX dialog there is no separate border delineating the chrome area from the client area on Windows 10.

And when the Mouse handling is removed on the following comment then clicking on a non-input area of the window causes it to 'jump' or temporarily resize since the unhandled Mouse click ripples up to the Window.  If you realy want this behavior can remove these handlers, but I think the existing behavior is preferred.  (And I know in my own mba when I displayed a read-only TextBox these Mouse handlers were necessary to prevent the unhandled exception crash that I indicated in the comment.  True we do not have a TextBox, but it seem to me like a valid defense position which also adds current UX value.   try it out and let me know what you want.

rseanhall replied:
In a standard WPF application, clicks on TextBoxes do not cause unhandled exceptions.  If the normal Appx installer doesn't have this drag behavior and we're adding click handlers to avoid crashing, then I think we should remove it.  If it does have the drag behavior, we need to find the root cause of the exception because I don't believe the click handlers are the correct way to fix it.

rseanhall commented on src/Setup/WixBA/RootView.xaml:
```diff
@@ -1,125 +1,487 @@
+            <!--  If click on no-input control (i.e. TextBlock, ProgressBar) catch click here to avoid InvalidOperationException.  -->
+            <DockPanel x:Name="AppArtBoardArea"
+                       VerticalAlignment="Stretch"
+                       MouseDown="RootView_IgnoreClick"
```
Since we don't have custom drag behavior, we shouldn't have to worry about any of these mouse events.

rseanhall commented on src/Setup/WixBA/RootView.xaml:
```diff
@@ -1,125 +1,487 @@
+                            <ColumnDefinition Width="5" />
+                        </Grid.ColumnDefinitions>
+
+                        <StackPanel x:Name="LeftMarqueeDcIsland"
```
I don't understand what the `DcIsland` suffix means.

phillHgl replied:
In general I find it useful to define the x:Name for most, or at least
major sections of XAML code, because when there are issues VS debugger uses
the x:Name in the Output messages.  The WixBA uses a WPF concept of Data
Context Islands (or DC Islands) where the DataContext property is set for a
group of controls, so that different behavior of a View is initialized with
possibly a different VM context than another set of controls.   When I was
first learning WPF and studying the WixBA I found the WixBA really
confusing because all the WPF purists were preaching having a single View
and ViewModel matched.  So when I understood the concept that the WixBA was
using of creating DataContext islands I found it useful to define x:Name to
make that intent clearer.  I also find it more useful to define x:Name
rather than using a comment, because comments cannot be nested, which gets
in the way of commenting out large blocks of XAML code to isolate errant
code.  (In my own projects I have implemented IoC and
PresentationTraceDebug handlers to catch these kinds of issues, but that
seemed like a more radical change than was proposed at this time.)

On Sun, Jul 31, 2016 at 11:54 PM, Sean Hall notifications@github.com
wrote:

> In src/Setup/WixBA/RootView.xaml
> https://github.com/wixtoolset/wix4/pull/208#discussion_r72924043:
> 
> > -                           Margin="0,0,10,10"
> > -                           HorizontalAlignment="Stretch"
> > -                           DockPanel.Dock="Bottom">
> > -                    <Grid x:Name="ActionGrid" HorizontalAlignment="Stretch">
> > -                        <Grid.RowDefinitions>
> > -                            <RowDefinition Height="auto" />
> > -                        </Grid.RowDefinitions>
> > -                        <Grid.ColumnDefinitions>
> > -                            <ColumnDefinition Width="5" />
> > -                            <ColumnDefinition Width="auto" />
> > -                            <ColumnDefinition Width="*" MinWidth="30" />
> > -                            <ColumnDefinition Width="auto" />
> > -                            <ColumnDefinition Width="5" />
> > -                        </Grid.ColumnDefinitions>
> >   +
> > -                        <StackPanel x:Name="LeftMarqueeDcIsland"
> 
> I don't understand what the DcIsland suffix means.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/wixtoolset/wix4/pull/208/files/28721ab8e4985a5641b3efd7975a67cecd8f8d4a#r72924043,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AHyCLQ0IYGS2OyHsgwUSS8fWCGTG42jmks5qbXwOgaJpZM4JVm8P
> .

rseanhall replied:
OK, just wondering.  I think the WPF purists would want each "DCIsland" to be a UserControl, which would result in each view with it's own view model.  I don't really think this UI is large enough for UserControls to be worth it.

rseanhall commented on src/Setup/WixBA/RootView.xaml:
```diff
@@ -1,125 +1,487 @@
+                            <StackPanel x:Name="UpdateBtnDcIsland"
+                                        DataContext="{Binding UpdateViewModel}"
+                                        Style="{DynamicResource ActionBtnStkPnlStyle}">
+                                <Button AutomationProperties.HelpText="An update is availalbe. Click to get the update"
```
Need to fix `availalbe` to `available`.

phillHgl replied:
Thanks!!

On Sun, Jul 31, 2016 at 11:56 PM, Sean Hall notifications@github.com
wrote:

> In src/Setup/WixBA/RootView.xaml
> https://github.com/wixtoolset/wix4/pull/208#discussion_r72924125:
> 
> > -                                             IsIndeterminate="{Binding CheckingEnabled}"
> > -                                             Style="{DynamicResource UpdateMarqueeStyle}"
> > -                                             Visibility="{Binding CheckingEnabled,
> > -                                                                  Converter={StaticResource BooleanToVisibilityConverter}}" />
> > -                            </StackPanel>
> > -                        </StackPanel>
> >   +
> > -                        <StackPanel x:Name="InitiateActionArea"
> > -                                    Grid.Column="3"
> > -                                    HorizontalAlignment="Right"
> > -                                    Orientation="Horizontal">
> > -                            <!--  Displayed from left to right, when ordered from top to bottom.  Most preferred is at the top.  -->
> > -                            <StackPanel x:Name="UpdateBtnDcIsland"
> > -                                        DataContext="{Binding UpdateViewModel}"
> > -                                        Style="{DynamicResource ActionBtnStkPnlStyle}">
> > -                                <Button AutomationProperties.HelpText="An update is availalbe. Click to get the update"
> 
> Need to fix availalbe to available.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/wixtoolset/wix4/pull/208/files/28721ab8e4985a5641b3efd7975a67cecd8f8d4a#r72924125,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AHyCLX6vSQWAjCOv66fKZd0LxWnXMFI6ks5qbXxigaJpZM4JVm8P
> .

rseanhall commented on src/Setup/WixBA/RootView.xaml:
```diff
@@ -1,125 +1,487 @@
+
+                <StackPanel x:Name="FinalStatusArea"
+                            HorizontalAlignment="Stretch"
+                            DockPanel.Dock="Bottom">
```
The sibling StackPanel above is already assigned to `DockPanel.Dock="Bottom"`.  Do not assign multiple controls the same docking location.  There are multiple ways to layout controls like this, my preference is to use Grids everywhere, I seldom use DockPanels or StackPanels.

phillHgl replied:
I agree that there are multiple ways to do layout in WPF.  DockPanel's do
not have to have a single DockPanel.Dock specification per syblings, and
the use of multiple docking is a common WPF practice that I think is
appropriate here.  I'm sure that it could be done with a Grid or some
other preferred approach depending on the reviewer.  I will ponder this,
since I mean to please.

On Sun, Jul 31, 2016 at 11:59 PM, Sean Hall notifications@github.com
wrote:

> In src/Setup/WixBA/RootView.xaml
> https://github.com/wixtoolset/wix4/pull/208#discussion_r72924244:
> 
> > -                                <Button AutomationProperties.HelpText="Press to Cancel"
> > -                                        AutomationProperties.Name="Cancel"
> > -                                        Command="{Binding CancelCommand}"
> > -                                        Content="Cancel"
> > -                                        Style="{DynamicResource FinalActionButtonStyle}"
> > -                                        Visibility="{Binding CancelEnabled,
> > -                                                             Converter={StaticResource BooleanToVisibilityConverter}}" />
> > -                            </StackPanel>
> >   +
> > -                        </StackPanel>
> > -                    </Grid>
> > -                </DockPanel>
> >   +
> > -                <StackPanel x:Name="FinalStatusArea"
> > -                            HorizontalAlignment="Stretch"
> > -                            DockPanel.Dock="Bottom">
> 
> The sibling StackPanel above is already assigned to
> DockPanel.Dock="Bottom". Do not assign multiple controls the same docking
> location. There are multiple ways to layout controls like this, my
> preference is to use Grids everywhere, I seldom use DockPanels or
> StackPanels.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/wixtoolset/wix4/pull/208/files/28721ab8e4985a5641b3efd7975a67cecd8f8d4a#r72924244,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AHyCLZ2R91vuoG3sEo8iSfYV8zWianEKks5qbX0XgaJpZM4JVm8P
> .

rseanhall replied:
The main thing I don't like here is that in order for this StackPanel to be above the previous StackPanel in the UI, it has to be below it in the XAML.

rseanhall commented on src/Setup/WixBA/RootView.xaml.cs:
```diff
@@ -26,6 +26,7 @@ public RootView(RootViewModel viewModel)
             this.InitializeComponent();
 
             viewModel.ViewWindowHandle = new WindowInteropHelper(this).EnsureHandle();
+            WindowProperties.Instance.CheckBackgroundBrightness();
```
We should be able to use this DependencyProperty declaratively.  Typically it would be a Boolean property, we set it to true in the xaml, and the property detects when it changes.  In the change handler, it would do `CheckBackgroundBrightness`.

rseanhall commented on src/Setup/WixBA/RootView.xaml.cs:
```diff
@@ -35,7 +36,37 @@ public RootView(RootViewModel viewModel)
+            RootViewModel rvm = this.DataContext as RootViewModel;
+            if ((null != rvm) && (InstallationState.Applying == rvm.InstallState))
+            {
+                rvm.Canceled = true;
```
I can't remember if WixBA prompts when the user clicks the Cancel button.  If so, we also should prompt here.  If not, there's probably a bug somewhere asking us to prompt.

phillHgl replied:
It does prompt on a Cancel.

On Mon, Aug 1, 2016 at 12:03 AM, Sean Hall notifications@github.com wrote:

> In src/Setup/WixBA/RootView.xaml.cs
> https://github.com/wixtoolset/wix4/pull/208#discussion_r72924455:
> 
> > -        private void RootView_IgnoreClick(object sender, MouseButtonEventArgs e)
> > -        {
> > -            //e.Handled = true;
> > -        }
> >   +
> > -        /// <summary>
> >   
> >   
> > -        /// Event is fired when the window is closing.
> > -        /// </summary>
> >   
> >   
> > -        /// <param name="sender"></param>
> > -        /// <param name="e"></param>
> > -        private void Window_Closing(object sender, System.ComponentModel.CancelEventArgs e)
> > -        {
> > -            RootViewModel rvm = this.DataContext as RootViewModel;
> > -            if ((null != rvm) && (InstallationState.Applying == rvm.InstallState))
> > -            {
> > -                rvm.Canceled = true;
> 
> I can't remember if WixBA prompts when the user clicks the Cancel button.
> If so, we also should prompt here. If not, there's probably a bug somewhere
> asking us to prompt.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/wixtoolset/wix4/pull/208/files/28721ab8e4985a5641b3efd7975a67cecd8f8d4a#r72924455,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AHyCLcB1zgGVc9a6I2LoopfxYA8oKv4vks5qbX4QgaJpZM4JVm8P
> .

rseanhall commented on src/Setup/WixBA/RootViewModel.cs:
```diff
@@ -43,6 +43,7 @@ public RootViewModel()
         public ProgressViewModel ProgressViewModel { get; private set; }
         public UpdateViewModel UpdateViewModel { get; private set; }
         public IntPtr ViewWindowHandle { get; set; }
+        public Action CloseAction { get; set; }
```
I can't find where this is used.

phillHgl replied:
True it was removed in a subsequent commit.

On Mon, Aug 1, 2016 at 12:03 AM, Sean Hall notifications@github.com wrote:

> In src/Setup/WixBA/RootViewModel.cs
> https://github.com/wixtoolset/wix4/pull/208#discussion_r72924466:
> 
> > @@ -43,6 +43,7 @@ public RootViewModel()
> >          public ProgressViewModel ProgressViewModel { get; private set; }
> >          public UpdateViewModel UpdateViewModel { get; private set; }
> >          public IntPtr ViewWindowHandle { get; set; }
> > -        public Action CloseAction { get; set; }
> 
> I can't find where this is used.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/wixtoolset/wix4/pull/208/files/28721ab8e4985a5641b3efd7975a67cecd8f8d4a#r72924466,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AHyCLdmpy9uDogfB6UDg0Tq-DL8B6xCWks5qbX4lgaJpZM4JVm8P
> .

phillHgl replied:
You are correct. Removed.

On Mon, Aug 1, 2016 at 12:03 AM, Sean Hall notifications@github.com wrote:

> In src/Setup/WixBA/RootViewModel.cs
> https://github.com/wixtoolset/wix4/pull/208#discussion_r72924466:
> 
> > @@ -43,6 +43,7 @@ public RootViewModel()
> >          public ProgressViewModel ProgressViewModel { get; private set; }
> >          public UpdateViewModel UpdateViewModel { get; private set; }
> >          public IntPtr ViewWindowHandle { get; set; }
> > -        public Action CloseAction { get; set; }
> 
> I can't find where this is used.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/wixtoolset/wix4/pull/208/files/28721ab8e4985a5641b3efd7975a67cecd8f8d4a#r72924466,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AHyCLdmpy9uDogfB6UDg0Tq-DL8B6xCWks5qbX4lgaJpZM4JVm8P
> .

rseanhall commented on src/Setup/WixBA/Styles.xaml:
```diff
@@ -1,103 +1,192 @@
 <?xml version="1.0" encoding="utf-8" ?>
-<!-- Copyright (c) .NET Foundation and contributors. All rights reserved. Licensed under the Microsoft Reciprocal License. See LICENSE.TXT file in the project root for full license information. -->
+<!--  Copyright (c) .NET Foundation and contributors. All rights reserved. Licensed under the Microsoft Reciprocal License. See LICENSE.TXT file in the project root for full license information.  -->
```
Extra spaces here, too.

phillHgl replied:
I think VS is automatically formatting two spaces at the beginning and end
of any comment, even if I remove the space.  I was not aware of any
WixToolset style for XAML.  I also have XamlStyler installed, but this
reformatting is happening even without invoking it.

https://visualstudiogallery.msdn.microsoft.com/d6634d0e-38fb-48b6-829f-dadbc5c2fb62

I am aware that I could go off and create custom rules, but since custom
rules for wix do not appear to have been specified, I am just trying to
adapt to using the defaults regardless of my personal preferences.

On Mon, Aug 1, 2016 at 12:04 AM, Sean Hall notifications@github.com wrote:

> In src/Setup/WixBA/Styles.xaml
> https://github.com/wixtoolset/wix4/pull/208#discussion_r72924491:
> 
> > @@ -1,103 +1,192 @@
> >  <?xml version="1.0" encoding="utf-8" ?>
> > -<!-- Copyright (c) .NET Foundation and contributors. All rights reserved. Licensed under the Microsoft Reciprocal License. See LICENSE.TXT file in the project root for full license information. -->
> > +<!--  Copyright (c) .NET Foundation and contributors. All rights reserved. Licensed under the Microsoft Reciprocal License. See LICENSE.TXT file in the project root for full license information.  -->
> 
> Extra spaces here, too.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/wixtoolset/wix4/pull/208/files/28721ab8e4985a5641b3efd7975a67cecd8f8d4a#r72924491,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AHyCLevCqe-d3b8aSRFwbUITc4CYnvQXks5qbX5MgaJpZM4JVm8P
> .

rseanhall replied:
OK, let's just stick with the defaults and not fight the IDE on this.

rseanhall commented on src/Setup/WixBA/WixBA.cs:
```diff
@@ -108,12 +142,22 @@ public static void PlanLayout()
         /// </summary>
         protected override void Run()
         {
+            // AttachDebuger();
```
Let's make sure to delete these when we're done updating the UI.

phillHgl replied:
I find this to be a useful tool now that the bundle creates a cr.  It certainly can be removed 'when we're done'.

rseanhall replied:
When I'm debugging a BA, I use Rob's suggestion of setting the `burn.clean.room` parameter to the path of the bundle so it doesn't actually create a clean room process.  That let me use my original process - set the BA project as the startup project and configure the project to launch the bundle for debugging.

rseanhall commented on src/Setup/WixBA/WixBA.csproj:
```diff
@@ -1,19 +1,48 @@
+  </PropertyGroup>
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Release|x86'">
+    <Prefer32Bit>false</Prefer32Bit>
+  </PropertyGroup>
```
These need to be deleted.  `Prefer32Bit` doesn't mean anything for DLLs.

phillHgl replied:
OK

On Mon, Aug 1, 2016 at 12:07 AM, Sean Hall notifications@github.com wrote:

> In src/Setup/WixBA/WixBA.csproj
> https://github.com/wixtoolset/wix4/pull/208#discussion_r72924638:
> 
> > -  </PropertyGroup>
> > -  <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Release|AnyCPU'">
> > -    <Prefer32Bit>false</Prefer32Bit>
> > -  </PropertyGroup>
> > -  <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Release|Win32'">
> > -    <Prefer32Bit>false</Prefer32Bit>
> > -  </PropertyGroup>
> > -  <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Release|arm'">
> > -    <Prefer32Bit>false</Prefer32Bit>
> > -  </PropertyGroup>
> > -  <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Release|x64'">
> > -    <Prefer32Bit>false</Prefer32Bit>
> > -  </PropertyGroup>
> > -  <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Release|x86'">
> > -    <Prefer32Bit>false</Prefer32Bit>
> > -  </PropertyGroup>
> 
> These need to be deleted. Prefer32Bit doesn't mean anything for DLLs.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/wixtoolset/wix4/pull/208/files/28721ab8e4985a5641b3efd7975a67cecd8f8d4a#r72924638,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AHyCLSbSisJ5rcGTI4Wjs3kUXQfTpV9uks5qbX8igaJpZM4JVm8P
> .

rseanhall commented on src/Setup/WixBA/WixBA.csproj:
```diff
@@ -31,36 +60,41 @@
+    </Reference>
     <Reference Include="System.ServiceModel" />
     <Reference Include="System.Windows.Forms" />
+    <Reference Include="System.Windows.Interactivity, Version=4.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
```
If we still need this, we might need to get it from a Nuget package.  I'm not sure if this comes with VS2010.

phillHgl commented:
I was initially confused by this as well until I realized that the same
behavior existed in the v3 WixBA and other WPF apps that I tried.  Since in
this app we are using the 'system' decoration of controls, rather than
offering our own styles reagarding focus, etc it looks to me like one needs
to click on the app somewhere before the system applies the adornment.  I
can keep researching this to see if I can understand it better and find a
solution.

On Sun, Jul 31, 2016 at 11:44 PM, Sean Hall notifications@github.com
wrote:

> This is looking really good.
> 
> When I launch it, the install button is disabled until I click somewhere.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/wixtoolset/wix4/pull/208#issuecomment-236488412, or mute
> the thread
> https://github.com/notifications/unsubscribe-auth/AHyCLSEyvvXtDGJHKvx9tWrH30yUuSRDks5qbXmngaJpZM4JVm8P
> .

rseanhall commented:
I wouldn't worry about it right now if that behavior was still there.  Maybe it wasn't as obvious in the old UI since the buttons were images.
